src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,358 @@
+/* 87858bc2-baa9-40b6-8e4c-2c95ba8b1545 */
+DECLARE_SOF_UUID("dp-schedule", dp_sched_uuid, 0x87858bc2, 0xbaa9, 0x40b6,
+		 0x8e, 0x4c, 0x2c, 0x95, 0xba, 0x8b, 0x15, 0x45);
+
+DECLARE_TR_CTX(dp_tr, SOF_UUID(dp_sched_uuid), LOG_LEVEL_INFO);
+
+struct scheduler_dp_data {
+	struct list_item tasks;		/* list of active dp tasks */
+	struct k_spinlock lock;		/* synchronization between cores */
+};
+
+struct task_dp_pdata {
+	k_tid_t thread_id;		/* zephyr thread ID */
+	k_thread_stack_t __sparse_cache *p_stack;	/* pointer to thread stack */
+	uint32_t ticks_period;		/* period the task should be scheduled in LL ticks */
+	uint32_t ticks_to_trigger;	/* number of ticks the task should be triggered after */
+	struct k_sem sem;		/* semaphore for task scheduling */
+};
+
+/*
+ * there's only one instance of DP scheduler for all cores
+ * Keep pointer to it here
+ */
+static struct scheduler_dp_data *dp_sch;
+
+static inline k_spinlock_key_t scheduler_dp_lock(void)
+{
+	return k_spin_lock(&dp_sch->lock);
+}
+
+static inline void scheduler_dp_unlock(k_spinlock_key_t key)
+{
+	k_spin_unlock(&dp_sch->lock, key);
+}
+
+/*
+ * function called after every LL tick
+ *
+ * TODO:
+ * the scheduler should here calculate deadlines of all task and tell Zephyr about them
+ * Currently there's an assumption that the task is always ready to run
+ */
+void scheduler_dp_ll_tick(void *receiver_data, enum notify_id event_type, void *caller_data)
+{
+	(void) receiver_data;
+	(void) event_type;
+	(void) caller_data;
+	struct list_item *tlist;
+	struct task *curr_task;
+	struct task_dp_pdata *pdata;
+	k_spinlock_key_t lock_key;
+
+	if (cpu_get_id() != PLATFORM_PRIMARY_CORE_ID)
+		return;
+
+	if (!dp_sch)
+		return;
+
+	lock_key = scheduler_dp_lock();
+	list_for_item(tlist, &dp_sch->tasks) {
+		curr_task = container_of(tlist, struct task, list);
+		pdata = curr_task->priv_data;
+
+		if (pdata->ticks_to_trigger == 0) {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED) {
+				/* set new trigger time, start the thread */
+				pdata->ticks_to_trigger = pdata->ticks_period;
+				curr_task->state = SOF_TASK_STATE_RUNNING;
+				k_sem_give(&pdata->sem);
+			}
+		} else {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED ||
+			    curr_task->state == SOF_TASK_STATE_RUNNING)
+				/* decrease num of ticks to re-schedule */
+				pdata->ticks_to_trigger--;
+		}
+	}
+	scheduler_dp_unlock(lock_key);
+}
+
+static int scheduler_dp_task_cancel(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_CANCEL;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	return 0;
+}
+
+static int scheduler_dp_task_free(void *data, struct task *task)
+{
+	k_spinlock_key_t lock_key;
+	struct task_dp_pdata *pdata = task->priv_data;
+
+	/* abort the execution of the thread */
+	k_thread_abort(pdata->thread_id);
```

Comment:
Thread abort is a little dangerous given the API here, where you have user callbacks that might hold locks or other resources at the point where they're aborted.  Ideally you'd want to do e.g. k_thread_join() here after having flagged a "terminate" command to the thread's main loop somehow.

Reply:
there's **scheduler_dp_task_cancel** for this - it will remove the thread from EDF scheduling and let it stop on main semaphore

scheduler_dp_task_free is a unconditional, async remove, it frees all the resources. We MUST be sure the thread is dead for good at this point. So -> abort is a good way to do it